apps/emqx/src/emqx_limiter/src/emqx_limiter_server.erl

```diff
@@ -0,0 +1,578 @@
+dispatch_burst_to_buckets(Childs, InFlow, Nodes) ->
+    Each = fun(ChildId, NodeAcc) ->
+                   #{counter := Counter,
+                     index := Index,
+                     obtained := Obtained} = Bucket = maps:get(ChildId, NodeAcc),
+                   counters:add(Counter, Index, InFlow),
+                   NodeAcc#{ChildId := Bucket#{obtained := Obtained + InFlow}}
+           end,
+    lists:foldl(Each, Nodes, Childs).
+
+-spec init_tree(emqx_limiter_schema:limiter_type(), state()) -> state().
+init_tree(Type, State) ->
+    #{global := Global,
+      zone := Zone,
+      bucket := Bucket} = emqx:get_config([emqx_limiter, Type]),
+    {Factor, Root} = make_root(Global, Zone),
+    State2 = State#{root := Root},
+    {NodeId, State3} = make_zone(maps:to_list(Zone), Factor, 1, State2),
+    State4 = State3#{counter := counters:new(maps:size(Bucket),
+                                             [write_concurrency])},
+    make_bucket(maps:to_list(Bucket), Global, Zone, Factor, NodeId, [], State4).
+
+-spec make_root(hocons:confg(), hocon:config()) -> {number(), root()}.
+make_root(#{rate := Rate, burst := Burst}, Zone) ->
+    ZoneNum = maps:size(Zone),
+    Childs = lists:seq(1, ZoneNum),
+    MiniPeriod = emqx_limiter_schema:minimum_period(),
+    if Rate >= 1 ->
+            {1, #{rate => Rate,
+                  burst => Burst,
+                  period => MiniPeriod,
+                  childs => Childs,
+                  consumed => 0}};
+       true ->
+            Factor = 1 / Rate,
+            {Factor, #{rate => 1,
+                       burst => Burst * Factor,
+                       period => erlang:floor(Factor * MiniPeriod),
+                       childs => Childs,
+                       consumed => 0}}
+    end.
+
+make_zone([{Name, ZoneCfg} | T], Factor, NodeId, State) ->
+    #{rate := Rate, burst := Burst} = ZoneCfg,
+    #{zones := Zones, nodes := Nodes} = State,
+    Zone = #{id => NodeId,
+             name => Name,
+             rate => mul(Rate, Factor),
+             burst => Burst,
+             obtained => 0,
+             childs => []},
+    State2 = State#{zones := Zones#{Name => NodeId},
+                    nodes := Nodes#{NodeId => Zone}},
+    make_zone(T, Factor, NodeId + 1, State2);
+
+make_zone([], _, NodeId, State2) ->
+    {NodeId, State2}.
+
+make_bucket([{Name, Conf} | T], Global, Zone, Factor, Id, Buckets, #{type := Type} = State) ->
+    #{zone := ZoneName,
+      aggregated := Aggregated} = Conf,
+    Path = emqx_limiter_manager:make_path(Type, ZoneName, Name),
+    case get_counter_rate(Conf, Zone, Global) of
+        infinite ->
+            State2 = State,
+            Rate = infinity,
+            Capacity = infinity,
+            Counter = undefined,
+            Index = undefined,
+            Ref = emqx_limiter_bucket_ref:new(Counter, Index, Rate),
+            emqx_limiter_manager:insert_bucket(Path, Ref);
+        #{rate := RawRate, capacity := Capacity} ->
```

Comment:
`get_counter_rate` returns the first limited node but neither Global nor Zone has `capacity` key thus could lead to crash here.